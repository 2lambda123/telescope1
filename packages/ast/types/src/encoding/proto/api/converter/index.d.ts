import * as t from '@babel/types';
import { ProtoType, ProtoRoot } from '@osmonauts/types';
import { ProtoParseContext } from '../../../context';
interface ApiConverterItemParams {
    root: ProtoRoot;
    context: ProtoParseContext;
    proto: ProtoType;
}
export declare const createApiConverterItem: ({ root, context, proto }: ApiConverterItemParams) => t.ObjectProperty;
interface ApiConverterParams {
    name: string;
    root: ProtoRoot;
    context: ProtoParseContext;
    protos: ProtoType[];
}
export declare const createApiConverter: ({ name, root, context, protos }: ApiConverterParams) => t.ExportNamedDeclaration;
export {};
