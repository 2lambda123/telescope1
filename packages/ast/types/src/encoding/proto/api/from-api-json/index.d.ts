import * as t from '@babel/types';
import { ProtoType, ProtoField } from '@osmonauts/types';
import { ProtoParseContext } from '../../../context';
export interface FromApiParseField {
    context: ProtoParseContext;
    field: ProtoField;
    currentProtoPath: string;
    scope: string[];
    fieldPath: ProtoField[];
    nested: number;
    isOptional: boolean;
}
export declare const fromApiParseField: ({ context, field, currentProtoPath, scope: previousScope, fieldPath: previousFieldPath, nested, isOptional }: FromApiParseField) => any;
interface FromApiJSON {
    context: ProtoParseContext;
    proto: ProtoType;
}
export declare const fromApiJsonInfo: ({ context, proto }: FromApiJSON) => {
    fromApiParams: t.ObjectPattern;
    blockStatement: t.BlockStatement;
    typeAnnotation: t.TSTypeAnnotation;
};
export declare const fromApiJsonMethod: ({ context, proto }: FromApiJSON) => t.ObjectMethod;
export declare const fromApiJsonFunction: ({ context, proto }: FromApiJSON) => t.ArrowFunctionExpression;
export {};
