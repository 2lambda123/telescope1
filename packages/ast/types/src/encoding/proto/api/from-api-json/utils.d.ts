import * as t from '@babel/types';
import { FromApiParseField } from './index';
export declare const fromApi: {
    defaultType(args: FromApiParseField): t.ObjectProperty;
    arrayFrom(args: FromApiParseField): t.ObjectProperty;
    type({ context, field, currentProtoPath, scope, fieldPath, nested, isOptional }: FromApiParseField): any;
    typeArray({ context, field, currentProtoPath, scope, fieldPath, nested, isOptional }: FromApiParseField): t.ObjectProperty;
    scalarArray({ context, field, currentProtoPath, scope, fieldPath, nested, isOptional }: FromApiParseField, arrayTypeAstFunc: Function): t.ObjectProperty;
};
