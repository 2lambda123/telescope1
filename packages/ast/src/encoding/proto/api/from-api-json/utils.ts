import * as t from '@babel/types';
import {
    BILLION,
    memberExpressionOrIdentifierApiCaseField,
    shorthandProperty
} from '../../../../utils';
import { FromApiParseField, fromApiParseField } from './index'
import { getFieldOptionality, getOneOfs } from '../../types';
import { ProtoField } from '@osmonauts/types';
import { protoFieldsToArray } from '../../../amino';

export const fromApi = {
    defaultType(args: FromApiParseField) {
        if (args.field.name === args.field?.options?.['(telescope:orig)']
            && args.scope.length === 1) {
            return shorthandProperty(args.field.name);
        }
        return t.objectProperty(
            t.identifier(args.field.name),
            memberExpressionOrIdentifierApiCaseField(args.fieldPath)
        );
    },

    arrayFrom(args: FromApiParseField) {
        return t.objectProperty(t.identifier(args.field.name),
            t.callExpression(
                t.memberExpression(
                    t.identifier('Array'),
                    t.identifier('from')
                ),
                [
                    memberExpressionOrIdentifierApiCaseField(args.fieldPath)
                ]
            ));
    },

    type({ context, field, currentProtoPath, scope, fieldPath, nested, isOptional }: FromApiParseField) {
        const parentField = field;
        const Type = context.getTypeFromCurrentPath(field, currentProtoPath);
        const oneOfs = getOneOfs(Type);
        const properties = protoFieldsToArray(Type).map(field => {
            const isOneOf = oneOfs.includes(field.name);
            const isOptional = getFieldOptionality(context, field, isOneOf);

            if (parentField.import) currentProtoPath = parentField.import;
            return fromApiParseField({
                context,
                field,
                currentProtoPath,
                scope: [...scope],
                fieldPath: [...fieldPath],
                nested: nested + 1,
                isOptional // TODO how to handle nested optionality?
            })
        });
        return t.objectProperty(t.identifier(field.name),
            t.objectExpression(
                properties
            )
        );
    },

    typeArray({ context, field, currentProtoPath, scope, fieldPath, nested, isOptional }: FromApiParseField) {
        //////
        const variable = 'el' + nested;
        const f = JSON.parse(JSON.stringify(field)); // clone
        const varProto: ProtoField = {
            ...f
        };
        varProto.name = variable;
        varProto.options['(telescope:orig)'] = variable;
        varProto.options['(telescope:name)'] = variable;
        varProto.options['(telescope:camel)'] = variable;
        //////

        const parentField = field;
        const Type = context.getTypeFromCurrentPath(field, currentProtoPath);
        const oneOfs = getOneOfs(Type);
        const properties = protoFieldsToArray(Type).map(field => {
            const isOneOf = oneOfs.includes(field.name);
            const isOptional = getFieldOptionality(context, field, isOneOf);

            if (parentField.import) currentProtoPath = parentField.import;

            return fromApiParseField({
                context,
                field,
                currentProtoPath,
                scope: [variable],
                fieldPath: [varProto],
                nested: nested + 1,
                isOptional // TODO how to handle nested optionality?
            })
        });

        const expr = t.callExpression(
            t.memberExpression(
                memberExpressionOrIdentifierApiCaseField(fieldPath),
                t.identifier('map')
            ),
            [
                t.arrowFunctionExpression(
                    [
                        t.identifier(variable)
                    ],
                    t.objectExpression(
                        properties
                    )
                )
            ]
        );

        return t.objectProperty(t.identifier(field.name),
            expr
        );
    },

    scalarArray({ context, field, currentProtoPath, scope, fieldPath, nested, isOptional }: FromApiParseField, arrayTypeAstFunc: Function) {
        const variable = 'el' + nested;

        const expr = t.callExpression(
            t.memberExpression(
                memberExpressionOrIdentifierApiCaseField(fieldPath),
                t.identifier('map')
            ),
            [
                t.arrowFunctionExpression(
                    [
                        t.identifier(variable)
                    ],
                    arrayTypeAstFunc(variable)
                )
            ]
        );

        return t.objectProperty(t.identifier(field.name),
            expr
        );
    }
};