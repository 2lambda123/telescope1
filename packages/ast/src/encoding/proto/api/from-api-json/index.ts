import * as t from '@babel/types';
import { ProtoType, ProtoField } from '@osmonauts/types';
import { arrowFunctionExpression, objectMethod } from '../../../../utils';
import { protoFieldsToArray } from '../../../amino';
import { fromApi } from './utils';
import { ProtoParseContext } from '../../../context';
import { getFieldOptionality, getOneOfs } from '../..';

const needsImplementation = (name: string, field: ProtoField) => {
    throw new Error(`need to implement fromApi (${field.type} rules[${field.rule}] name[${name}])`);
}

export interface FromApiParseField {
    context: ProtoParseContext;
    field: ProtoField;
    currentProtoPath: string;
    scope: string[];
    fieldPath: ProtoField[];
    nested: number;
    isOptional: boolean;
};

export const fromApiParseField = ({
    context,
    field,
    currentProtoPath,
    scope: previousScope,
    fieldPath: previousFieldPath,
    nested,
    isOptional
}: FromApiParseField) => {

    const scope = [field.name, ...previousScope];
    const fieldPath = [field, ...previousFieldPath];

    const args = {
        context,
        field,
        currentProtoPath,
        scope,
        fieldPath,
        nested,
        isOptional
    };

    // arrays
    if (field.rule === 'repeated') {
        switch (field.type) {
            case 'string':
            case 'int64':
            case 'sint64':
            case 'uint64':
            case 'fixed64':
            case 'sfixed64':
            case 'double':
            case 'float':
            case 'int32':
            case 'sint32':
            case 'uint32':
            case 'fixed32':
            case 'sfixed32':
            case 'bool':
            case 'bytes':
                return fromApi.defaultType(args);
        }

        switch (field.parsedType.type) {
            case 'Type':
                return fromApi.typeArray(args);
            case 'Enum':
            case 'cosmos.base.v1beta1.Coin':
                return fromApi.defaultType(args);
        }

        return needsImplementation(field.name, field);
    }

    // special types...
    switch (field.type) {
        case 'Timestamp':
        case 'google.protobuf.Timestamp':
        case 'ibc.core.client.v1.Height':
        case 'Height':
        case 'Duration':
        case 'google.protobuf.Duration':
            return fromApi.defaultType(args)

        default:
    }

    // Types/Enums
    switch (field.parsedType.type) {
        case 'Type':
            return fromApi.type(args);

        case 'Enum':
            return fromApi.defaultType(args);
    }

    // scalar types...
    switch (field.type) {
        case 'string':
        case 'int64':
        case 'sint64':
        case 'uint64':
        case 'fixed64':
        case 'sfixed64':
        case 'double':
        case 'float':
        case 'int32':
        case 'sint32':
        case 'uint32':
        case 'fixed32':
        case 'sfixed32':
        case 'bool':
        case 'bytes':
        default:
            return fromApi.defaultType(args)
    }
};

interface FromApiJSON {
    context: ProtoParseContext;
    proto: ProtoType;
}

export const fromApiJsonInfo = ({
    context,
    proto
}: FromApiJSON) => {

    const fromApiParams = t.objectPattern(
        Object.keys(proto.fields).map((field) => t.objectProperty(
            t.identifier(proto.fields[field].options['(telescope:orig)']),
            t.identifier(proto.fields[field].options['(telescope:orig)']),
            false,
            true)
        )
    );
    fromApiParams.typeAnnotation = t.tsTypeAnnotation(t.tsIndexedAccessType(
        t.tsTypeReference(t.identifier('Api' + proto.name)),
        t.tsLiteralType(t.stringLiteral('value'))
    ));

    const oneOfs = getOneOfs(proto);
    const fields = protoFieldsToArray(proto).map((field) => {
        const isOneOf = oneOfs.includes(field.name);
        const isOptional = getFieldOptionality(context, field, isOneOf);

        const apiField = fromApiParseField({
            context,
            field,
            currentProtoPath: context.ref.filename,
            scope: [],
            fieldPath: [],
            nested: 0,
            isOptional
        });
        return {
            ctx: context,
            field: apiField
        }
    });

    return {
        fromApiParams,
        blockStatement: t.blockStatement([
            t.returnStatement(
                t.objectExpression(
                    fields.map(({ field }) => field)
                )
            )
        ]),
        typeAnnotation: t.tsTypeAnnotation(t.tsTypeReference(t.identifier(proto.name)))
    };

};

export const fromApiJsonMethod = ({
    context,
    proto
}: FromApiJSON) => {

    const {
        fromApiParams,
        blockStatement,
        typeAnnotation
    } = fromApiJsonInfo({
        context,
        proto
    });

    return objectMethod(
        'method',
        t.identifier('fromApi'),
        [
            fromApiParams
        ],
        blockStatement,
        false,
        false,
        false,
        typeAnnotation
    );
};


export const fromApiJsonFunction = ({
    context,
    proto
}: FromApiJSON) => {

    const {
        fromApiParams,
        blockStatement,
        typeAnnotation
    } = fromApiJsonInfo({
        context,
        proto
    });

    return arrowFunctionExpression(
        [
            fromApiParams
        ],
        blockStatement,
        typeAnnotation
    );
};

