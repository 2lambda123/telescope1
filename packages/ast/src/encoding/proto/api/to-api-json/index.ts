import * as t from '@babel/types';
import { arrowFunctionExpression, objectMethod } from '../../../../utils';
import { ProtoParseContext } from '../../../context';
import { ProtoType, ProtoField } from '@osmonauts/types';
import { getFieldOptionality, getOneOfs } from '../..';
import { protoFieldsToArray } from '../../../amino';
import { toApi } from './utils';

const needsImplementation = (name: string, field: ProtoField) => {
    throw new Error(`need to implement toApi (${field.type} rules[${field.rule}] name[${name}])`);
}

const warningDefaultImplementation = (name: string, field: ProtoField) => {
    console.warn(`need to implement toApi (${field.type} rules[${field.rule}] name[${name}])`);
}

export interface ToApiParseField {
    context: ProtoParseContext;
    field: ProtoField;
    currentProtoPath: string;
    scope: string[];
    fieldPath: ProtoField[];
    nested: number;
    isOptional: boolean;
};

export const toApiParseField = ({
    context,
    field,
    currentProtoPath,
    scope: previousScope,
    fieldPath: previousFieldPath,
    nested,
    isOptional
}: ToApiParseField) => {

    const scope = [field.name, ...previousScope];
    const fieldPath = [field, ...previousFieldPath];

    const args = {
        context,
        field,
        currentProtoPath,
        scope,
        fieldPath,
        nested,
        isOptional
    };

    // arrays
    if (field.rule === 'repeated') {
        switch (field.type) {
            case 'string':
            case 'int64':
            case 'sint64':
            case 'uint64':
            case 'fixed64':
            case 'sfixed64':
            case 'double':
            case 'float':
            case 'int32':
            case 'sint32':
            case 'uint32':
            case 'fixed32':
            case 'sfixed32':
            case 'bool':
            case 'bytes':
            case 'string':
                return toApi.defaultType(args);
        }

        switch (field.parsedType.type) {
            case 'Enum':
                return toApi.defaultType(args);
            case 'Type':
                return toApi.defaultType(args);
                if (fieldPath.length > 10) {
                    return toApi.defaultType(args);
                }
                return toApi.typeArray(args);
        }

        return needsImplementation(field.name, field);
    }

    // special types...
    switch (field.type) {
        case 'Timestamp':
        case 'google.protobuf.Timestamp':
        case 'cosmos.base.v1beta1.Coin':
        case 'ibc.core.client.v1.Height':
        case 'Height':
        case 'Duration':
        case 'google.protobuf.Duration':
            return toApi.defaultType(args)
        default:
    }

    // Types/Enums
    switch (field.parsedType.type) {
        case 'Enum':
            return toApi.defaultType(args);
        case 'Type':
            return toApi.defaultType(args);
            if (fieldPath.length > 10) {
                return toApi.defaultType(args);
            }
            return toApi.type(args);
    }

    // scalar types...
    switch (field.type) {
        case 'string':
        case 'int64':
        case 'sint64':
        case 'uint64':
        case 'fixed64':
        case 'sfixed64':
        case 'double':
        case 'float':
        case 'int32':
        case 'sint32':
        case 'uint32':
        case 'fixed32':
        case 'sfixed32':
        case 'bool':
        case 'bytes':
            return toApi.defaultType(args)

        default:
            warningDefaultImplementation(field.name, field);
            return toApi.defaultType(args)
    }
};

export const toApiJsonInfo = (
    context: ProtoParseContext,
    name: string,
    proto: ProtoType
) => {

    const toApiParams = t.objectPattern(
        protoFieldsToArray(proto).map((field) =>
            t.objectProperty(
                t.identifier(field.name),
                t.identifier(field.name),
                false,
                true)
        )
    );

    toApiParams.typeAnnotation = t.tsTypeAnnotation(t.tsTypeReference(t.identifier(name)))

    const oneOfs = getOneOfs(proto);

    const fields = protoFieldsToArray(proto).map((field) => {
        const isOneOf = oneOfs.includes(field.name);
        const isOptional = getFieldOptionality(context, field, isOneOf);

        const apiField = toApiParseField({
            context,
            field,
            currentProtoPath: context.ref.filename,
            scope: [],
            fieldPath: [],
            nested: 0,
            isOptional
        });
        return {
            ctx: context,
            field: apiField
        }
    });

    return {
        toApiParams,
        blockStatement: t.blockStatement([
            t.returnStatement(
                t.objectExpression(
                    fields.map(({ field }) => field)
                )
            )
        ]),
        typeAnnotation: t.tsTypeAnnotation(t.tsIndexedAccessType(
            t.tsTypeReference(t.identifier('Api' + name)),
            t.tsLiteralType(t.stringLiteral('value'))
        ))
    };
};

export const toApiJsonMethod = (
    context: ProtoParseContext,
    name: string,
    proto: ProtoType
) => {

    const {
        toApiParams,
        blockStatement,
        typeAnnotation
    } = toApiJsonInfo(
        context,
        name,
        proto
    );

    return objectMethod(
        'method',
        t.identifier('toApi'),
        [
            toApiParams
        ],
        blockStatement,
        false,
        false,
        false,
        typeAnnotation
    );
};

export const toApiJsonFunction = (
    context: ProtoParseContext,
    name: string,
    proto: ProtoType
) => {

    const {
        toApiParams,
        blockStatement,
        typeAnnotation
    } = toApiJsonInfo(
        context,
        name,
        proto
    );

    return arrowFunctionExpression(
        [
            toApiParams
        ],
        blockStatement,
        typeAnnotation
    );
};