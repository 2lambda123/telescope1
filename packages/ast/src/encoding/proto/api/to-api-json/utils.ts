import * as t from '@babel/types';
import { memberExpressionOrIdentifier, shorthandProperty } from "../../../../utils";
import { ToApiParseField, toApiParseField } from './index'
import { ProtoField } from '@osmonauts/types';
import { protoFieldsToArray } from '../../../amino';
import { getFieldOptionality, getOneOfs } from '../../types';

export const toApi = {
    defaultType(args: ToApiParseField) {
        if (args.field.name === args.field.options['(telescope:orig)'] && args.scope.length === 1) {
            return shorthandProperty(args.field.name);
        }
        return t.objectProperty(t.identifier(args.field.options['(telescope:orig)']), memberExpressionOrIdentifier(args.scope))
    },

    type({ context, field, currentProtoPath, scope, fieldPath, nested, isOptional }: ToApiParseField) {
        /// TODO (can this be refactored out? e.g. no recursive calls in this file?)
        /// BEGIN
        const Type = context.getTypeFromCurrentPath(field, currentProtoPath);
        const parentField = field;
        const oneOfs = getOneOfs(Type);
        const properties = protoFieldsToArray(Type).map(field => {
            const isOneOf = oneOfs.includes(field.name);
            const isOptional = getFieldOptionality(context, field, isOneOf);

            if (parentField.import) currentProtoPath = parentField.import;

            return toApiParseField({
                context,
                field,
                currentProtoPath,
                scope: [...scope],
                fieldPath: [...fieldPath],
                nested,
                isOptional // TODO how to handle nested optionality
            })
        });
        /// END 
        return t.objectProperty(t.identifier(field.options['(telescope:orig)']),
            t.objectExpression(
                properties
            )
        );
    },

    typeArray({ context, field, currentProtoPath, scope, fieldPath, nested, isOptional }: ToApiParseField) {
        //////
        const variable = 'el' + nested;
        const f = JSON.parse(JSON.stringify(field)); // clone
        const varProto: ProtoField = {
            ...f
        };
        varProto.name = variable;
        varProto.options['(telescope:orig)'] = variable;
        varProto.options['(telescope:name)'] = variable;
        varProto.options['(telescope:camel)'] = variable;
        //////

        if (field.parsedType.type !== 'Type') {
            throw new Error('Arrays only support types[Type] right now.');
        }

        const Type = context.getTypeFromCurrentPath(field, currentProtoPath);
        const parentField = field;
        const oneOfs = getOneOfs(Type);

        const properties = protoFieldsToArray(Type).map(field => {
            const isOneOf = oneOfs.includes(field.name);
            const isOptional = getFieldOptionality(context, field, isOneOf);

            if (parentField.import) currentProtoPath = parentField.import;

            if (!field.name) {
                console.log('missing name:')
                console.log(JSON.stringify(field, null, 2))
            }

            return toApiParseField({
                context,
                field,
                currentProtoPath,
                scope: [variable],
                fieldPath: [varProto],
                nested: nested + 1,
                isOptional // TODO how to handle nested optionality
            });
        });


        const expr = t.callExpression(
            t.memberExpression(
                memberExpressionOrIdentifier(scope),
                t.identifier('map')
            ),
            [
                t.arrowFunctionExpression(
                    [
                        t.identifier(variable)
                    ],
                    t.objectExpression(
                        properties
                    )
                )
            ]
        );

        return t.objectProperty(t.identifier(field.options['(telescope:orig)']),
            expr
        );
    },

    scalarArray({ context, field, currentProtoPath, scope, nested, isOptional }: ToApiParseField, arrayTypeAstFunc: Function) {
        const variable = 'el' + nested;

        const expr = t.callExpression(
            t.memberExpression(
                memberExpressionOrIdentifier(scope),
                t.identifier('map')
            ),
            [
                t.arrowFunctionExpression(
                    [
                        t.identifier(variable)
                    ],
                    arrayTypeAstFunc(variable)
                )
            ]
        );

        return t.objectProperty(t.identifier(field.options['(telescope:orig)']),
            expr
        );
    }
};
