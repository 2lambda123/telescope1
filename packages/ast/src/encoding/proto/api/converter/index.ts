import * as t from '@babel/types';
import { ProtoType, ProtoRoot } from '@osmonauts/types';
import { toApiJsonFunction } from '../to-api-json';
import { ProtoParseContext } from '../../../context';
import { getTypeUrl } from '../../../amino';
import { fromApiJsonFunction } from '../from-api-json';

interface AminoConverterItemParams {
    root: ProtoRoot,
    context: ProtoParseContext,
    proto: ProtoType
}
export const createApiConverterItem = (
    {
        root,
        context,
        proto
    }: AminoConverterItemParams
) => {

    const typeUrl = getTypeUrl(root, proto);

    return t.objectProperty(
        t.stringLiteral(typeUrl),
        t.objectExpression(
            [
                t.objectProperty(
                    t.identifier('toApi'),
                    toApiJsonFunction({
                        context,
                        proto
                    })
                ),
                t.objectProperty(
                    t.identifier('fromApi'),
                    fromApiJsonFunction({
                        context,
                        proto
                    })
                )
            ]
        )
    );
};




interface ApiConverterParams {
    name: string,
    root: ProtoRoot,
    context: ProtoParseContext,
    protos: ProtoType[]
}
export const createApiConverter = (
    {
        name,
        root,
        context,
        protos
    }: ApiConverterParams) => {

    const items = protos.map(proto => {
        return createApiConverterItem({
            context,
            root,
            proto
        })
    })

    return t.exportNamedDeclaration(t.variableDeclaration('const', [
        t.variableDeclarator(t.identifier(name),
            t.objectExpression(
                items
            ))
    ]));
};

