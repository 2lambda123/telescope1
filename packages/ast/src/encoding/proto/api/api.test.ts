import { makeApiInterface } from './index';
import { ProtoStore } from '@osmonauts/proto-parser'
import { snake } from 'case';
import { prepareContext, expectCode } from '../../../../test-utils';
import { fromApiJsonMethod, fromApiParseField } from './from-api-json';
import { toApiJsonMethod } from './to-api-json';
import { createApiConverter } from './converter';
const store = new ProtoStore([__dirname + '/../../../../../../__fixtures__/chain1']);
store.traverseAll();

describe('cosmos/staking/v1beta1/tx', () => {
    const {
        context, protos, root
    } = prepareContext(store, 'cosmos/staking/v1beta1/tx.proto')

    it('MsgCreateValidator', () => {
        context.options.aminoEncoding.casingFn = snake;
        expectCode(makeApiInterface(
            {
                context,
                proto: protos.find(p => p.name === 'MsgCreateValidator'),
            }
        ))
    })
    it('from-api', () => {
        context.options.aminoEncoding.casingFn = snake;
        expectCode(fromApiJsonMethod(
            {
                context,
                proto: protos.find(p => p.name === 'MsgCreateValidator'),
            }
        ))
    })
    it('to-api', () => {
        context.options.aminoEncoding.casingFn = snake;
        expectCode(toApiJsonMethod(
            {
                context,
                proto: protos.find(p => p.name === 'MsgCreateValidator'),
            }
        ))
    })
    it('converter', () => {
        context.options.aminoEncoding.casingFn = snake;
        expectCode(createApiConverter(
            {
                context,
                root,
                name: 'ApiConverter',
                protos: protos.filter(p => p.name === 'MsgCreateValidator'),
            }
        ))
    })
});

