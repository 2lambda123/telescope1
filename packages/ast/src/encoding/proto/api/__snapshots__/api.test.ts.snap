// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cosmos/staking/v1beta1/tx MsgCreateValidator 1`] = `
"export interface ApiMsgCreateValidator {
  type: \\"cosmos-sdk/MsgCreateValidator\\";
  value: {
    description: {
      moniker: string;
      identity: string;
      website: string;
      security_contact: string;
      details: string;
    };
    commission: {
      rate: string;
      max_rate: string;
      max_change_rate: string;
    };
    min_self_delegation: string;
    delegator_address: string;
    validator_address: string;
    pubkey: {
      type_url: string;
      value: Uint8Array;
    };
    value: {
      denom: string;
      amount: string;
    };
  };
}"
`;

exports[`cosmos/staking/v1beta1/tx converter 1`] = `
"export const ApiConverter = {
  \\"/cosmos.staking.v1beta1.MsgCreateValidator\\": {
    toApi: ({
      description,
      commission,
      minSelfDelegation,
      delegatorAddress,
      validatorAddress,
      pubkey,
      value
    }: MsgCreateValidator): ApiMsgCreateValidator[\\"value\\"] => {
      return {
        description: {
          moniker: description.moniker,
          identity: description.identity,
          website: description.website,
          security_contact: description.securityContact,
          details: description.details
        },
        commission: {
          rate: commission.rate,
          max_rate: commission.maxRate,
          max_change_rate: commission.maxChangeRate
        },
        min_self_delegation: minSelfDelegation,
        delegator_address: delegatorAddress,
        validator_address: validatorAddress,
        pubkey: {
          type_url: pubkey.typeUrl,
          value: pubkey.value
        },
        value
      };
    },
    fromApi: ({
      description,
      commission,
      min_self_delegation,
      delegator_address,
      validator_address,
      pubkey,
      value
    }: ApiMsgCreateValidator[\\"value\\"]): MsgCreateValidator => {
      return {
        description: {
          moniker: description.moniker,
          identity: description.identity,
          website: description.website,
          securityContact: description.security_contact,
          details: description.details
        },
        commission: {
          rate: commission.rate,
          maxRate: commission.max_rate,
          maxChangeRate: commission.max_change_rate
        },
        minSelfDelegation: min_self_delegation,
        delegatorAddress: delegator_address,
        validatorAddress: validator_address,
        pubkey: {
          typeUrl: pubkey.type_url,
          value: pubkey.value
        },
        value: {
          denom: value.denom,
          amount: value.amount
        }
      };
    }
  }
};"
`;

exports[`cosmos/staking/v1beta1/tx from-api 1`] = `
"fromApi({
  description,
  commission,
  min_self_delegation,
  delegator_address,
  validator_address,
  pubkey,
  value
}: ApiMsgCreateValidator[\\"value\\"]): MsgCreateValidator {
  return {
    description: {
      moniker: description.moniker,
      identity: description.identity,
      website: description.website,
      securityContact: description.security_contact,
      details: description.details
    },
    commission: {
      rate: commission.rate,
      maxRate: commission.max_rate,
      maxChangeRate: commission.max_change_rate
    },
    minSelfDelegation: min_self_delegation,
    delegatorAddress: delegator_address,
    validatorAddress: validator_address,
    pubkey: {
      typeUrl: pubkey.type_url,
      value: pubkey.value
    },
    value: {
      denom: value.denom,
      amount: value.amount
    }
  };
}"
`;

exports[`cosmos/staking/v1beta1/tx to-api 1`] = `
"toApi({
  description,
  commission,
  minSelfDelegation,
  delegatorAddress,
  validatorAddress,
  pubkey,
  value
}: MsgCreateValidator): ApiMsgCreateValidator[\\"value\\"] {
  return {
    description: {
      moniker: description.moniker,
      identity: description.identity,
      website: description.website,
      security_contact: description.securityContact,
      details: description.details
    },
    commission: {
      rate: commission.rate,
      max_rate: commission.maxRate,
      max_change_rate: commission.maxChangeRate
    },
    min_self_delegation: minSelfDelegation,
    delegator_address: delegatorAddress,
    validator_address: validatorAddress,
    pubkey: {
      type_url: pubkey.typeUrl,
      value: pubkey.value
    },
    value
  };
}"
`;
